std/assembly/array.ts

```diff
@@ -493,6 +493,68 @@ export class Array<T> {
     return <string>unreachable();
   }
 
+  flat(): T {
+    if (!isArray<T>()) {
+      ERROR("Cannot call flat() on Array<T> where T is not an Array.");
+    }
+    // Get the length and data start values
+    var length = this.length_;
+    var selfDataStart = this.dataStart;
+
+    // calculate the end size with an initial pass
+    var size = 0;
+    for (let i = 0; i < length; i++) {
+      let child = load<usize>(selfDataStart + (i << alignof<T>()));
+      size += child == 0 ? 0 : load<i32>(child, offsetof<T>("length_"));
+    }
+
+    // calculate the byteLength of the resulting backing ArrayBuffer
+    var byteLength = <usize>size << usize(alignof<valueof<T>>());
+    var dataStart = __alloc(byteLength, idof<ArrayBuffer>());
+
+    // create the return value and initialize it
+    var result = __alloc(offsetof<T>(), idof<T>());
+    store<i32>(result, size, offsetof<T>("length_"));
+
+    // byteLength, dataStart, and buffer are all readonly
+    store<i32>(changetype<usize>(result), byteLength, offsetof<T>("byteLength"));
+    store<usize>(changetype<usize>(result), dataStart, offsetof<T>("dataStart"));
+    store<usize>(changetype<usize>(result), __retain(dataStart), offsetof<T>("buffer"));
+
+    // set the elements
+    var resultIndex = -1;
+    for (let i = 0; i < length; i++) { // for each child
+      let child = load<usize>(selfDataStart + (i << alignof<T>()));
+
+      // ignore null arrays
+      if (child == 0) continue;
+
+      // move to the next array index
+      resultIndex++;
+
+      // copy the underlying buffer data to the result buffer
+      let childDataLength = load<i32>(child, offsetof<T>("byteLength"));
+      memory.copy(
+        dataStart + (<usize>resultIndex << usize(alignof<valueof<T>>())),
+        load<usize>(child, offsetof<T>("dataStart")),
+        <usize>childDataLength,
+      );
+
+      // advance the result length
+      let childLength = childDataLength >> <i32>alignof<valueof<T>>();
+      resultIndex += childLength - 1;
+    }
```

Comment:
it look like you could simplify this:
```ts
var resultIndex = 0;
for (let i = 0; i < length; i++) {
   ...
   // resultIndex++; don't need this

   resultIndex += childLength;
}
  ...
}
```

Reply:
actually, that part was required by the algorithm. I made a compromise and refactored the algorithm. Please check it again.